plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseIcebergConnectorTest.java

```diff
@@ -4488,6 +4483,34 @@ public void testOptimizeCleansUpDeleteFiles()
         assertUpdate("DROP TABLE " + tableName);
     }
 
+    @Test
+    public void testOptimizeSnapshot()
+    {
+        Session session = Session.builder(getSession())
+                .setCatalogSessionProperty(getSession().getCatalog().orElseThrow(), "allow_legacy_snapshot_syntax", "true")
+                .build();
+        String tableName = "test_optimize_snapshot_" + randomTableSuffix();
+
+        assertUpdate("CREATE TABLE " + tableName + " (a) AS VALUES 11", 1);
+        long snapshotId = getCurrentSnapshotId(tableName);
+        assertUpdate("INSERT INTO " + tableName + " VALUES 22", 1);
+        assertThatThrownBy(() -> query(session, "ALTER TABLE \"%s@%d\" EXECUTE OPTIMIZE".formatted(tableName, snapshotId)))
+                .hasMessage("Cannot execute table procedure OPTIMIZE on old snapshot " + snapshotId);
```

Comment:
nit: How about `assertQueryFails` to ensure `TrinoException`?

Reply:
`assertThatThrownBy(() -> query(...))` is a pattern we use very widely.
if we want to deprecate it in favor of `assertQueryFails`, we should have a discussion about this first.